sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java

```diff
@@ -311,4 +311,49 @@ default boolean purgeTable(Identifier ident) throws UnsupportedOperationExceptio
    */
   void renameTable(Identifier oldIdent, Identifier newIdent)
       throws NoSuchTableException, TableAlreadyExistsException;
+
+  /**
+   * Instantiate a builder to create a table in the catalog.
+   *
+   * @param ident  a table identifier.
+   * @param columns the columns of the new table.
+   * @return the TableBuilder to create a table.
+   */
+  default TableBuilder buildTable(Identifier ident, Column[] columns) {
+    return new TableBuilderImpl(this, ident, columns);
+  }
+
+  /**
+   * Builder used to create tables.
+   *
+   * <p>Call {@link #buildTable(Identifier, Column[])} to create a new builder.
+   */
+  interface TableBuilder {
+    /**
+     * Sets the partitions for the table.
+     *
+     * @param partitions Partitions for the table.
+     * @return this for method chaining
+     */
+    TableBuilder withPartitions(Transform[] partitions);
+
+    /**
+     * Adds key/value properties to the table.
+     *
+     * @param properties key/value properties
+     * @return this for method chaining
+     */
+    TableBuilder withProperties(Map<String, String> properties);
```

Comment:
It seems  'withProperties' sounds like it should set/replace the existing map, rather than increment the existing map, based on how the other method is defined.

In Spark the builder arguments for conf seem to take in single key, value.  I wonder if we should follow this pattern if the goal is to have an additive API , else it seems wasteful (callers have to make a temporary java Map just to call this).

Maybe we can support both APIs?

Just my observation.

Reply:
Maybe we should add extra method. e.g. `TableBuilder addProperty(String key, String val);`.

Reply:
I am inclined towards just making the API not additive. Most call sites already have the table properties already materialized as a map. Updating the PR and the Javadoc accordingly. Please let me know if that sounds reasonable.

Reply:
We can add new methods like `addProperty` once there is a use case for it. I do agree with Szehon it should replace the original map, like the PR currently does.

Reply:
Personally, I guess `TableBuilder addProperty(String key, String val);` will be used when updating table.

Reply:
+1 with @szehon-ho @aokolnychyi @beliefer . We can explicitly mention that this API is for set/replace the whole properties in code comments.